kin-core/src/main/java/kin/core/PaymentWatcher.java

```diff
@@ -0,0 +1,114 @@
+package kin.core;
+
+
+import static kin.core.Utils.checkNotNull;
+
+import android.support.annotation.NonNull;
+import java.math.BigDecimal;
+import java.util.List;
+import kin.core.ServiceProvider.KinAsset;
+import org.stellar.sdk.KeyPair;
+import org.stellar.sdk.Memo;
+import org.stellar.sdk.MemoHash;
+import org.stellar.sdk.Operation;
+import org.stellar.sdk.PaymentOperation;
+import org.stellar.sdk.Server;
+import org.stellar.sdk.requests.EventListener;
+import org.stellar.sdk.responses.TransactionResponse;
+import org.stellar.sdk.sse.ServerSentEvent;
+
+/**
+ * Watch blockchain network for ongoing payments involved specific account.
+ * <p>Use {@link #start(WatcherListener)} to start
+ * watching, new payments will be notified on the input {@link WatcherListener WatcherListener&lt;PaymentInfo&gt;}</p>
+ */
+public class PaymentWatcher {
+
+    private static final String CURSOR_FUTURE_ONLY = "now";
+    private final Server server;
+    private final KinAsset kinAsset;
+    private final KeyPair accountKeyPair;
+    private boolean started;
+    private ServerSentEvent serverSentEvent;
+
+    PaymentWatcher(Server server, Account account, KinAsset kinAsset) {
+        this.server = server;
+        this.kinAsset = kinAsset;
+        accountKeyPair = KeyPair.fromAccountId(account.getAccountId());
+    }
+
+    /**
+     * Start watching for payments, use {@link #stop()} to stop watching.
+     * <p><b>Note:</b> Events will be fired on background thread.</p>
+     *
+     * @param listener listener object for payment events
+     */
+    public void start(@NonNull final WatcherListener<PaymentInfo> listener) {
+        checkNotNull(listener, "listener");
+        if (started) {
+            throw new IllegalStateException("Watcher already started");
+        }
+        started = true;
+        serverSentEvent = server
+            .transactions()
```

Comment:
I think you can listen to payment operation instead https://www.stellar.org/developers/horizon/reference/endpoints/payments-for-account.html
and then you will need to filter less

Reply:
if it include the memo as well?

Reply:
you're right, it was considered, but payment does not give us all the required info, the memo exist  only at transaction level